src/fabric.c

```diff
@@ -80,15 +83,25 @@ static int fi_register_provider(struct fi_provider *provider)
 
 	prov = fi_getprov(provider->name);
 	if (prov) {
-		/* If we have two versions of the same provider,
-		 * keep the most recent
+		/* If this provider is older than an already-loaded
+		 * provider of the same name, then discard this one.
 		 */
 		if (FI_VERSION_GE(prov->provider->version, provider->version)) {
 			ret = -FI_EALREADY;
 			goto cleanup;
 		}
 
+		/* This provider is newer than an already-loaded
+		 * provider of the same name, so discard the
+		 * already-loaded one.
+		 */
 		prov->provider->cleanup();
+#ifdef HAVE_LIBDL
+		if (prov->dlhandle)
+			dlclose(prov->dlhandle);
+#endif
```

Comment:
I may be being dense here -- but why don't we need #ifdef?  We need to protect the use of dlclose, right?